Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -77,43 +78,49 @@ instance : AddMonoid DyckWord where
 
 namespace DyckWord
 
-variable (p q : DyckWord)
+variable {p q : DyckWord}
 
 lemma toList_eq_nil : p.toList = [] ↔ p = 0 := by rw [DyckWord.ext_iff]; rfl
+lemma toList_ne_nil : p.toList ≠ [] ↔ p ≠ 0 := toList_eq_nil.ne
+
+variable (h : p ≠ 0)
 
 /-- The first element of a nonempty Dyck word is `U`. -/
-lemma head_eq_U (h : p.toList ≠ []) : p.toList.head h = U := by
+lemma head_eq_U : p.toList.head (toList_ne_nil.mpr h) = U := by
```

Comment:
Having `h : p ≠ 0` instead of `h : p.toList ≠ []` means that you will need to provide the hypothesis when rewriting. In those cases where an hypothesis appears in the LHS, we pretend it's not an argument to the lemma and make other arguments explicit in consequence. Here this gives
```suggestion
lemma head_eq_U (p : DyckWord) (h) : p.toList.head h = U := by
```
Note that the `(h)` is an extra sign that Lean can infer `h` from the LHS